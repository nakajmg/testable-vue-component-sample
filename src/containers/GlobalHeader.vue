<template>
  <header class="Header">
    <Logo class="Header_Logo" />
    <span class="Header_SiteTitle"
      @click="navigateRoot"
    >
      <SiteTitle :title="siteTitle"/>
    </span>
    <Menu class="Header_Menu"
      :items="menuItems"
      @clickMenuItem="onClickMenuItem"
    />
  </header>
</template>

<script>
import { mapState } from "vuex"
import Logo from "../basics/Logo.vue"
import SiteTitle from "../basics/SiteTitle.vue"
import Menu from "../components/Menu.vue"
export default {
  name: "Header",
  components: {
    Logo,
    SiteTitle,
    Menu,
  },
  computed: {
    ...mapState(["siteTitle", "menuItems"]),
  },
  methods: {
    onClickMenuItem({ name }) {
      this.$emit("navigate", { name })
    },
    navigateRoot() {
      this.$emit("navigate", { name: "root" })
    },
  },
}
</script>

<style lang="scss">
.Header {
  padding: 0.5em 1em;
  background-color: #3eaf7c;
  display: flex;
  align-items: center;
  &_Logo {
    width: 35px;
  }
  &_SiteTitle {
    color: #fff;
    font-size: 1.75em;
    margin-left: 0.5em;
    cursor: pointer;
  }
  &_Menu {
    color: #fff;
    font-size: 1.25em;
    margin-left: auto;
  }
}
</style>
