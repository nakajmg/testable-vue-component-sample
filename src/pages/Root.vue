<template>
  <div class="Root">
    <GlobalHeader @navigate="onNavigate"/>
    <Logo class="Root_Logo"/>
  </div>
</template>

<script>
import GlobalHeader from "../containers/GlobalHeader.vue"
import Logo from "../basics/Logo.vue"
export default {
  name: "Root",
  components: {
    GlobalHeader,
    Logo,
  },
  methods: {
    onNavigate({ name }) {
      this.$router.push({ name })
    },
  },
}
</script>

<style lang="scss">
.Root {
  display: flex;
  flex-direction: column;
  height: 100vh;
  &_Logo {
    width: 40%;
    max-width: 400px;
    margin: 0 auto;
    flex-grow: 1;
    display: inline-flex;
    align-items: center;
  }
}
</style>
