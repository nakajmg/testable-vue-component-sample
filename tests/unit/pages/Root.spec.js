import { createLocalVue, shallowMount } from "@vue/test-utils"
import Vuex from "vuex"
import menuItems from "../_mockData/menuItems.json"
import Root from "@/pages/Root.vue"
import GlobalHeader from "@/containers/GlobalHeader.vue"
const localVue = createLocalVue()
localVue.use(Vuex)
describe("Root.vue", () => {
  let store
  let $router
  beforeAll(() => {
    store = new Vuex.Store({
      state: {
        siteTitle: "test site title",
        menuItems,
      },
    })
    // $routerのモックを作成
    $router = {
      push: jest.fn(), // $router.pushをモック関数にしておく
    }
  })
  describe("methos", () => {
    it("onNavigate", () => {
      // Vueインスタンス作成時に$routerプロパティを注入する
      const wrapper = shallowMount(Root, { store, localVue, mocks: { $router } })
      wrapper.vm.onNavigate({ name: "root" })
      // mock化した$router.pushが呼ばれているか
      expect($router.push).toHaveBeenCalledWith({
        name: "root",
      })
    })
  })
  describe("template", () => {
    it("@navigate=onNavigate", () => {
      const mock = jest.fn()
      const wrapper = shallowMount(Root, { store, localVue, mocks: { $router } })
      wrapper.setMethods({
        onNavigate: mock,
      })
      const globalHeader = wrapper.find(GlobalHeader)
      // "navigate"イベントをemitする
      globalHeader.vm.$emit("navigate")
      expect(mock).toHaveBeenCalled()
    })
    it("snapshot", () => {
      const wrapper = shallowMount(Root, { store, localVue, mocks: { $router } })
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})
