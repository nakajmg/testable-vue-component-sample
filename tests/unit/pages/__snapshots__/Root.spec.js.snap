// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root.vue template snapshot 1`] = `
<div
  class="Root"
>
  <header
    class="Header"
  >
    <div
      class="Logo Header_Logo"
    >
      <img
        alt="Vue logo"
        class="Logo_Img"
        src="../assets/logo.png"
      />
    </div>
     
    <span
      class="Header_SiteTitle"
    >
      <div
        class="SiteTitle"
      >
        
  test site title

      </div>
    </span>
     
    <div
      class="Menu Header_Menu"
    >
      <div
        class="MenuItem Menu_Item"
      >
        <span
          class="MenuItem_Label"
        >
          
    About
  
        </span>
      </div>
      <div
        class="MenuItem Menu_Item"
      >
        <span
          class="MenuItem_Label"
        >
          
    Todos
  
        </span>
      </div>
    </div>
  </header>
   
  <div
    class="Logo Root_Logo"
  >
    <img
      alt="Vue logo"
      class="Logo_Img"
      src="../assets/logo.png"
    />
  </div>
</div>
`;
